### ISSUE TITLE
docker stack deploy creates dangling images

### ISSUE THREAD

**Issue by yaskoo** [docker/docker#31252]
**Description**

Deploying a stack with a compose file creates dangling images if the images listed in the compose file aren't already pulled.

**Steps to reproduce the issue:**
1. sample compose.yaml
```
version: '3'
services:
    redis:
        image: redis
```
2. run ```docker stack deploy --compose-file compose.yaml mystack```

**Describe the results you received:**
After the stack is deployed running ```docker images --filter="dangling=true"```, reports that redis is dangling

**Describe the results you expected:**
redis to be properly tagged

**Additional information you deem important (e.g. issue happens only occasionally):**
I remember reading somewhere that docker in swarm mode will use the image digests, but couldn't they be tagged as well?

**Output of `docker version`:**

```
Client:
 Version:      1.13.1
 API version:  1.26
 Go version:   go1.7.5
 Git commit:   092cba3
 Built:        Wed Feb  8 06:42:29 2017
 OS/Arch:      linux/amd64

Server:
 Version:      1.13.1
 API version:  1.26 (minimum version 1.12)
 Go version:   go1.7.5
 Git commit:   092cba3
 Built:        Wed Feb  8 06:42:29 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

**Output of `docker info`:**

```
Containers: 14
 Running: 9
 Paused: 0
 Stopped: 5
Images: 9
Server Version: 1.13.1
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 129
 Dirperm1 Supported: false
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins: 
 Volume: local
 Network: bridge host macvlan null overlay
Swarm: active
 NodeID: q9c6v9br4eijyzhoicrfa7hbd
 Is Manager: true
 ClusterID: 4efgqqf4tuv99a0bk79b6zcxv
 Managers: 1
 Nodes: 1
 Orchestration:
  Task History Retention Limit: 5
 Raft:
  Snapshot Interval: 10000
  Number of Old Snapshots to Retain: 0
  Heartbeat Tick: 1
  Election Tick: 3
 Dispatcher:
  Heartbeat Period: 5 seconds
 CA Configuration:
  Expiry Duration: 3 months
 Node Address: 172.17.8.100
 Manager Addresses:
  172.17.8.100:2377
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: aa8187dbd3b7ad67d8e5e3a15115d3eef43a7ed1
runc version: 9df8b306d01f59d3a8029be411de015b7304dd8f
init version: 949e6fa
Security Options:
 apparmor
Kernel Version: 3.13.0-109-generic
Operating System: Ubuntu 14.04.5 LTS
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 7.798 GiB
Name: *hostname*
ID: GSIT:B5UZ:64MD:6E5H:7DGY:GAD5:YDFQ:RW6A:3Z6X:O3Y7:TTVE:FYWO
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Comment by thaJeztah** [comment]
When deploying a service, docker resolves the digest for the image you specify, and "pins" the service to that digest. Doing so makes sure that every instance of the service runs _exactly_ the same version of the image. This prevents situations where, for example `redis:latest` on "node A" is an older version of the `:latest` image than `redis:latest` on "node B".

Given that the service won't be referring to the `redis:latest` image, but to `redis@sha256:aabbbccc...`, the `redis:latest` image will be seen as "dangling". Given that (if the `redis:latest` image was up-to-date), the `redis:latest` is the same image as `redis@sha256:aab..`, removing that dangling image will only "untag" the image, but not remove the actual image layers (which are still referenced by `redis@sha256:aab..`)

**Comment by thaJeztah** [comment]
Oh, wait, I misunderstood; I see the problem; it doesn't see `redis@sha:foo` as "in use", and prints it as "dangling".

That looks like a bug indeed

**Comment by yaskoo** [comment]
Exactly
btw your previous message is exactly what I remember reading ...thanks for writing it again

**Comment by yaskoo** [comment]
Just to clarify a bit, it seems that the engine actually uses redis@foo and this only the docker images command

**Comment by aaronlehmann** [comment]
I think this is a bit confusing, but it's ultimately not a big problem.

The image is technically a dangling image because there's no tag pointing to it.

If a running container is using that image, it won't be removed by `docker image prune` (at least it shouldn't be).

If there's no running container, it could be removed by `docker image prune` or through manual housekeeping, but it would be pulled the next time swarm starts a container using the image, so no harm would be done.

**Comment by nishanttotla** [comment]
I believe this is how distribution works right now. Does this only happen
when the compose file doesn't specify a tag?

In the case of no tag, latest is assumed by default, and the image pinned
to digest, but the downloaded image isn't tagged "latest". I suppose this
could be fixed, but I'm not sure of why it is like this.

@aaronlehmann or @dmcgowan might know. I can work on a PR it if we agree on
the fix.

On Wed, Feb 22, 2017 at 03:52 Yasen Terziivanov <notifications@github.com>
wrote:

Just to clarify a bit, it seems that the engine actually uses redis@foo and
this only the docker images command

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/docker/docker/issues/31252#issuecomment-281648570>, or mute
the thread

**Comment by aaronlehmann** [comment]
We don't tag the image because that might overwrite a local tag.

**Comment by yaskoo** [comment]
@nishanttotla 

> Does this only happen when the compose file doesn't specify a tag?

No, it happens with any tag, and yes it's not a big deal; stuff works - it was just confusing to me and just looks ~broken

**Comment by foleymic** [comment]
The only issue I see with these images being reported as dangling is that when you do a docker system prune, they will be removed - is that the desired behavior?

**Comment by thaJeztah** [comment]
@foleymic but they won't be removed if there's still a container using it (or are you seeing that happening?)

**Comment by foleymic** [comment]
@thaJeztah - Correct, I am not seeing that, prune is only removing Images that don't have running containers.  My issue is that due to #32620 we have to perform a docker system prune on each node after executing docker stack rm.  Then every time we start the stacks back up, all the images need to be pulled again because the prune removed them because it sees all these images as dangling.

**Comment by yaskoo** [comment]
> but they won't be removed if there's still a container using it

yes, they will not be removed, but the prune command will fail ...probably half way (until it reached one of the "dangling" images)
